linker/shared/src/main/scala/org/scalajs/linker/backend/BasicLinkerBackend.scala

```diff
@@ -98,8 +101,15 @@ final class BasicLinkerBackend(config: LinkerBackendImpl.Config)
           jsFileWriter.write(printedModuleSetCache.headerBytes)
           jsFileWriter.writeASCIIString("'use strict';\n")
 
-          for (printedTree <- printedTrees)
-            jsFileWriter.write(printedTree.jsCode)
+          if (config.minify) {
+            val printer = new Printers.JSTreePrinter(jsFileWriter)
+            for (tree <- trees)
+              printer.printStat(tree)
+          } else {
+            val printedTrees = (trees: List[js.Tree]).asInstanceOf[List[js.PrintedTree]]
+            for (printedTree <- printedTrees)
+              jsFileWriter.write(printedTree.jsCode)
+          }
```

Comment:
Does it make sense to abstract over this with an additional class so we can avoid the cast to `List[js.PrintedTree]` with a path-dependent type)?

Like:

```scala
abstract class BodyPrinter {
  type Tree
  
  val postTransformer: PostTransformer[Tree]

  def printWithoutSourceMap(trees: List[Tree], jsFileWriter: ByteArrayWriter): Unit
  def printWithSourceMap(/* snip */)
}
```

Reply:
I added a SQUASH commit that implements this suggestion. It is more verbose, but it does get rid of the cast. Let me know if it looks good to you.

Reply:
I think we should keep it yes. It's a bit more verbose, but IMO the call flow is actually easier to trace / reason about.